### PR TITLE
module parsing: make heuristic to get paths from modules more robust

### LINKED ISSUE #5273: [Question] External packages module requirement to work with Spack?
Suppose I have following external module:

```
module show nix/bgq/cmake/3.3.2
-------------------------------------------------------------------
/gpfs-path-xxx//modules-bgq-lugano-all/nix/bgq/cmake/3.3.2:

module-whatis	 cmake for BGQ backend  generated by nix
prepend-path	 PATH /gpfs-path/store/636l7bvg95fvdyffddw8g94cxwsrb1v8-generated-env-module-cmake/bin
-------------------------------------------------------------------
```

Then I add entry in `packages.yaml` as :

```
    cmake:
        modules:
            cmake@3.3.2%xl@12.1 arch=bgq-cnk-ppc64: nix/bgq/cmake/3.3.2
        buildable: False
        version: [3.3.2]
```

While installation I see:

```
==> cmake@3.9.0 : generating module file
==> Error: 'NoneType' object has no attribute 'startswith'
```

Debug shows :

```
==> cmake@3.3.2 : generating module file
==> SKIP: shebang filtering [external package]
==> READ LOCK: /gpfs/bbp.cscs.ch/home/kumbhar/spack_install/bgq/.spack-db/lock[0:0] [Acquiring]
==> READ LOCK: /gpfs/bbp.cscs.ch/home/kumbhar/spack_install/bgq/.spack-db/lock[0:0] [Acquired]
==> READ LOCK: /gpfs/bbp.cscs.ch/home/kumbhar/spack_install/bgq/.spack-db/lock[0:0] [Released]
==> 	WRITE : cmake@3.3.2%xl@12.1~doc+ncurses+openssl+ownlibs~qt arch=bgq-cnk-ppc64 /gshkzty [/gpfs/bbp.cscs.ch/home/kumbhar/spack_modules/bgq/tcl/bgq-cnk-ppc64/cmake-3.3.2-xl-12.1-gshkzty]
Traceback (most recent call last):
  File "/gpfs-path/softwares/sources/spack/bin/spack", line 81, in <module>
    sys.exit(spack.main.main())
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/main.py", line 552, in main
    return _main(command, parser, args, unknown)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/main.py", line 444, in _main
    return _invoke_spack_command(command, parser, args, unknown_args)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/main.py", line 355, in _invoke_spack_command
    return_val = command(parser, args)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/cmd/install.py", line 361, in install
    package.do_install(**kwargs)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/package.py", line 1272, in do_install
    **kwargs
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/package.py", line 1272, in do_install
    **kwargs
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/package.py", line 1233, in do_install
    return self._process_external_package(explicit)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/package.py", line 1175, in _process_external_package
    spack.hooks.post_install(self.spec)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/hooks/__init__.py", line 73, in __call__
    hook(*args, **kwargs)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/hooks/module_file_generation.py", line 32, in post_install
    generator.write()
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/modules.py", line 380, in write
    env = inspect_path(self.spec.prefix)
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/spec.py", line 1364, in prefix
    return Prefix(spack.store.layout.path_for_spec(self))
  File "/gpfs-path/softwares/sources/spack/lib/spack/spack/directory_layout.py", line 127, in path_for_spec
    assert(not path.startswith(self.root))
AttributeError: 'NoneType' object has no attribute 'startswith'
```

So are there any prerequisite for external modules to play nicely with Spack?

cc: @alalazo

### LINKED ISSUE #9126: failing to use already existing module hdf5
Can't give all details because of environment of operaiton.

==> hdf5@1.8.16 : has external module in cray-hdf5-parallel/1.8.16
==> hdf5@1.8.16 : is actually installed in None
==> hdf5@1.8.16 : generating module file
Traceback (most recent call last):
  File "/home/svadlam/spack-sv/bin/spack", line 54, in <module>
    sys.exit(spack.main.main())
  File "/home/svadlam/spack-sv/lib/spack/spack/main.py", line 653, in main
    return _invoke_command(command, parser, args, unknown)
  File "/home/svadlam/spack-sv/lib/spack/spack/main.py", line 432, in _invoke_command
    return_val = command(parser, args)
  File "/home/svadlam/spack-sv/lib/spack/spack/cmd/install.py", line 273, in install
    install_spec(args, kwargs, spec)
  File "/home/svadlam/spack-sv/lib/spack/spack/cmd/install.py", line 154, in install_spec
    spec.package.do_install(**kwargs)
  File "/home/svadlam/spack-sv/lib/spack/spack/package.py", line 1420, in do_install
    **kwargs)
  File "/home/svadlam/spack-sv/lib/spack/spack/package.py", line 1380, in do_install
    return self._process_external_package(explicit)
  File "/home/svadlam/spack-sv/lib/spack/spack/package.py", line 1302, in _process_external_package
    spack.hooks.post_install(self.spec)
  File "/home/svadlam/spack-sv/lib/spack/spack/hooks/__init__.py", line 73, in __call__
    hook(*args, **kwargs)
  File "/home/svadlam/spack-sv/lib/spack/spack/hooks/module_file_generation.py", line 48, in <lambda>
    post_install = lambda spec: _for_each_enabled(spec, 'write')
  File "/home/svadlam/spack-sv/lib/spack/spack/hooks/module_file_generation.py", line 41, in _for_each_enabled
    getattr(generator, method_name)()
  File "/home/svadlam/spack-sv/lib/spack/spack/modules/common.py", line 698, in write
    context = self.context.to_dict()
  File "/home/svadlam/spack-sv/lib/spack/spack/tengine.py", line 86, in to_dict
    d = [(name, getattr(self, name)) for name in self.context_properties]
  File "/home/svadlam/spack-sv/lib/spack/spack/modules/common.py", line 523, in environment_modifications
    self.spec.prefix,
  File "/home/svadlam/spack-sv/lib/spack/spack/spec.py", line 1406, in prefix
    self.prefix = spack.store.layout.path_for_spec(self)
  File "/home/svadlam/spack-sv/lib/spack/spack/directory_layout.py", line 97, in path_for_spec
    assert(not path.startswith(self.root))
AttributeError: 'NoneType' object has no attribute 'startswith'

---

### Steps to reproduce the issue

```console
$  spack --debug -k install --log-file log.txt --dirty -j 4 -v nalu-wind %gcc@7.2.0 '^trilinos~alloptpkgs~xsdkflags~metis~mumps~superlu-dist+superlu~hypre+hdf5~suite-sparse~python~shared+boost+tpetra~epetra~epetraext+exodus+pnetcdf+zlib+stk+teuchos+belos+zoltan+zoltan2~amesos+amesos2~anasazi~ifpack+ifpack2+muelu~fortran~ml+gtest~aztec~sacado~x11+instantiate~instantiate_cmplx~dtk~fortrilinos+openmp~rol~nox+shards~intrepid~intrepid2~cgns@develop'
... # and the relevant part of the error message
```

### Platform and user environment

Please report your OS here:
```commandline
$ uname -a 
Linux mutrino 4.4.114-94.11-default #1 SMP Thu Feb 1 19:28:26 UTC 2018 (4309ff9) x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	SUSE Linux Enterprise Server 12 SP3
``` 

I've seen another issue 
https://github.com/spack/spack/issues/5273#issuecomment-328363658
but no resolution in that question helped with this issue.

### LINKED ISSUE #9710: Spack not looking in correct PATH when multiple PATHs used in modulefile
When pointing Spack to a modulefile that includes multiple instances of `PATH`, the behavior of Spack appears to be to stop at the first instance of `PATH`. This can lead to errors when using Spack to install packages that depend on this modulefile.

Example modulefile contents:

```
/usr/common/software/modulefiles/python/2.7-anaconda-5.2:

conflict	 python 
module-whatis	 Python programming language.
unsetenv	 PYTHONSTARTUP 
setenv		 PYTHONUSERBASE /global/homes/r/rowlandk/.local/cori/2.7-anaconda-5.2 
prepend-path	 PATH /global/common/cori/software/python/2.7-anaconda-5.2/lib/python2.7/site-packages/mpi4py/bin
prepend-path	 PATH /global/common/cori/software/python/2.7-anaconda-5.2/bin 
```
With this, performing a Spack installation with the Python module as a dependency gives:

```
==> Error: RuntimeError: Unable to locate python command in /global/common/cori/software/python/2.7-anaconda-5.2/lib/python2.7/site-packages/mpi4py/bin
```

The second instance of `PATH` in the modulefile is what ends up at the beginning of the user's `PATH` upon loading the module, but the way that Spack reads in the modulefile and takes the first instance of `PATH` leads to this erroneous behavior.

---

We have worked around this issue by changing relevant modulefiles, but I would think that Spack's behavior should read all of the `PATH` in the order resultant from loading the module.

### LINKED ISSUE #12650: external packages: spack finds wrong prefix for cuda from module on Cori
I'm seeing a problem where spack finds the wrong prefix for a cuda
module.  IMO, this has to be either a bug with the module entry or
else a bug in how spack reads the module, and I'm asking for a ruling.

This is on cori at NERSC, a Cray machine with GPUs on some compute
nodes.  I created a dummy package that just adds a few dependencies
and has a trivial install() method that prints the `spec[...].prefix`
for the packages.

```
class Mypack(Package):
    ...
    depends_on('cmake')
    depends_on('cuda')

    def install(self, spec, prefix):
        print 'cmake =  {0}'.format(spec['cmake'].prefix)
        print 'cuda =   {0}'.format(spec['cuda'].prefix)
	...
```

I added `modules:` entries in `packages.yaml` for cmake, cuda, etc.

```
cmake:
  modules:
    cmake@3.14.4:  cmake/3.14.4

cuda:
  modules:
    cuda@10.1.168:  cuda/10.1.168
```

Most of the prefixes make sense, but the cuda prefix is `/usr` which
is not correct.

```
cmake =  /global/common/sw/cray/cnl7/haswell/cmake/3.14.4/gcc/8.2.0/2hef55n
cuda =   /usr
jdk =    /global/common/cori_cle7/software/jdk/1.8.0_202
papi =   /opt/cray/pe/papi/5.7.0.1
```

I'm expecting `spec['cuda'].prefix` to be `/usr/common/software/cuda/10.1.168`,
not `/usr`.

Cori uses TCL modules and the cuda module has:

```
$ module show cuda/10.1.168
-------------------------------------------------------------------
/usr/common/software/modulefiles/cuda/10.1.168:

conflict         cuda 
module-whatis    CUDA is a parallel computing platform and programming model developed by NVIDIA
                 for general computing on graphical processing unit (GPU). 
setenv           CUDA_VERSION 10.1.168 
setenv           CUDNN_VERSION 7.6.2 
setenv           CUDA_ROOT /usr/common/software/cuda/10.1.168 
setenv           CUDA_HOME /usr/common/software/cuda/10.1.168 
setenv           CUDA_PATH /usr/common/software/cuda/10.1.168 
prepend-path     LD_LIBRARY_PATH /usr/lib64 
prepend-path     LD_LIBRARY_PATH /usr/common/software/cuda/10.1.168/lib64 
prepend-path     PATH /usr/common/software/cuda/10.1.168/bin 
prepend-path     PATH /usr/common/software/cuda/10.1.168/NsightCompute-2019.3.1.2 
prepend-path     PATH /usr/common/software/cuda/10.1.168/NsightSystems-linux-public-2019.3.6.30-e2e891f/Host-x86_64/ 
prepend-path     PATH /usr/common/software/cuda/10.1.168/NsightSystems-linux-public-2019.3.6.30-e2e891f/Target-x86_64/x86_64 
prepend-path     MANPATH /usr/common/software/cuda/10.1.168/doc/man 
prepend-path     CMAKE_PREFIX_PATH /usr/common/software/cuda/10.1.168 
-------------------------------------------------------------------
```

It's not TCL modules in general.  The other modules on cori are TCL
and they seem to work.

It's not just this one cuda module.  There are 5 cuda modules on cori,
older and newer.  The others also report /usr.

It's not cuda in general.  I tried other cuda modules on other
systems, including summit at ORNL, and they work.  But those are LMOD
modules.

It seems to be just cuda on cori at NERSC.  At least, I haven't found
another pattern.

Cori has separate FE/BE arch types.  Same problem for either arch.

I'm not blocked.  I can change `modules:` to `paths:` and get the
right prefix.  But this is supposed to work, right?

This is with a recent spack clone: 02bb1d5ec618 on Aug 27.

Anyway, I'm seeking a ruling if this a problem with the module entry
or in how spack reads the module.  But let's try not to have each side
blame the other.  :-)

ping @mamelara who seems to be in the intersection of spack and NERSC.

### LINKED ISSUE #12662: external package: path extracted from module does not exist and contains colons.
I'm getting a weird warning message with the `intel-parallel-studio` module on my cluster:
```
==> Warning: Extracted path from module does not exist:
	Extracted path: /usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/intel64/slib:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/compiler
```
However, both of those directories do exist:
```console
$ ls /usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/intel64/slib
libdwarf.so  libelf.so  libVTcs.so  libVTfs.so  libVTim.so  libVTmc.so  libVTnull.so  libVT.so  libvtunwind.so  Makefile  slib_common  VT_sample.c
$ ls /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/compiler
include  lib  perf_headers
```
The module file contains:
```console
$ module show intel
-------------------------------------------------------------------
/usr/local/modulefiles/intel/18.0:

module-whatis	 Name:        ICC 
module-whatis	 Long Name:   Parallel Studio XE 2018 Cluster Edition update 1 
module-whatis	 Version:     2018, update 1 
module-whatis	 Category:    compiler 
module-whatis	 Description: The GNU compiler 
setenv		 VT_MPI impi4 
setenv		 INTEL_PYTHONHOME /usr/local/intel/parallel_studio_xe_2018/debugger_2018/python/intel64/ 
setenv		 DAALROOT /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/daal 
prepend-path	 CPLUS_INCLUDE_PATH /usr/local/intel/parallel_studio_xe_2018/clck/2018.1/include 
prepend-path	 PATH /usr/local/intel/parallel_studio_xe_2018/advisor_2018.1.1.535164/bin64:/usr/local/intel/parallel_studio_xe_2018/vtune_amplifier_2018.1.0.535340/bin64:/usr/local/intel/parallel_studio_xe_2018/inspector_2018.1.1.535159/bin64:/usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/intel64/bin:/usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/intel64/bin:/usr/local/intel/parallel_studio_xe_2018/clck/2018.1/bin/intel64:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/bin/intel64:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mpi/intel64/bin 
append-path	 PATH /usr/local/intel/parallel_studio_xe_2018/parallel_studio_xe_2018.1.038/bin 
setenv		 CLCK_ROOT /usr/local/intel/parallel_studio_xe_2018/clck/2018.1 
prepend-path	 MANPATH /usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/man:/usr/local/intel/parallel_studio_xe_2018/man/common:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mpi/man:/usr/local/intel/parallel_studio_xe_2018/documentation_2018/en/debugger/gdb-ia/man:/usr/local/intel/parallel_studio_xe_2018/documentation_2018/en/debugger/gdb-igfx/man 
setenv		 MPM_LAUNCHER /usr/local/intel/parallel_studio_xe_2018/debugger_2018/mpm/mic/bin/start_mpm.sh 
prepend-path	 LIBRARY_PATH /usr/local/intel/parallel_studio_xe_2018/clck/2018.1/lib/intel64:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/ipp/lib/intel64:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/compiler/lib/intel64_lin:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mkl/lib/intel64_lin:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/tbb/lib/intel64/gcc4.4:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/tbb/lib/intel64/gcc4.4:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/daal/lib/intel64_lin 
setenv		 ADVISOR_2018_DIR /usr/local/intel/parallel_studio_xe_2018/advisor_2018.1.1.535164 
setenv		 VT_ROOT /usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017 
prepend-path	 CLASSPATH /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mpi/intel64/lib/mpi.jar:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/daal/lib/daal.jar 
setenv		 PSTLROOT /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/pstl 
setenv		 I_MPI_ROOT /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mpi 
setenv		 GDB_CROSS /usr/local/intel/parallel_studio_xe_2018/debugger_2018/gdb/intel64/bin/gdb-ia 
setenv		 IPPROOT /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/ipp 
setenv		 INSPECTOR_2018_DIR /usr/local/intel/parallel_studio_xe_2018/inspector_2018.1.1.535159 
setenv		 VT_ARCH intel64 
prepend-path	 INTEL_LICENSE_FILE 1700@cc-lic1.campuscluster.illinois.edu 
prepend-path	 CPATH /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/ipp/include:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mkl/include:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/pstl/include:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/tbb/include:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/tbb/include:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/daal/include 
setenv		 VT_LIB_DIR /usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/intel64/lib 
setenv		 VT_ADD_LIBS -ldwarf -lelf -lvtunwind -lm -lpthread 
prepend-path	 LD_LIBRARY_PATH /usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/intel64/slib:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/compiler/lib/intel64:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/compiler/lib/intel64_lin:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mpi/intel64/lib:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mpi/mic/lib:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/ipp/lib/intel64:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/compiler/lib/intel64_lin:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mkl/lib/intel64_lin:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/tbb/lib/intel64/gcc4.4:/usr/local/intel/parallel_studio_xe_2018/debugger_2018/iga/lib:/usr/local/intel/parallel_studio_xe_2018/debugger_2018/libipt/intel64/lib:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/daal/lib/intel64_lin 
setenv		 TBBROOT /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/tbb 
setenv		 VT_SLIB_DIR /usr/local/intel/parallel_studio_xe_2018/itac/2018.1.017/intel64/slib 
setenv		 VTUNE_AMPLIFIER_2018_DIR /usr/local/intel/parallel_studio_xe_2018/vtune_amplifier_2018.1.0.535340 
prepend-path	 INFOPATH /usr/local/intel/parallel_studio_xe_2018/documentation_2018/en/debugger/gdb-ia/info:/usr/local/intel/parallel_studio_xe_2018/documentation_2018/en/debugger/gdb-igfx/info/ 
setenv		 GDBSERVER_MIC /usr/local/intel/parallel_studio_xe_2018/debugger_2018/gdb/targets/intel64/x200/bin/gdbserver 
setenv		 MKLROOT /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mkl 
prepend-path	 PYTHONPATH /usr/local/intel/parallel_studio_xe_2018/advisor_2018.1.1.535164/pythonapi 
prepend-path	 NLSPATH /usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/compiler/lib/intel64/locale/%l_%t/%N:/usr/local/intel/parallel_studio_xe_2018/compilers_and_libraries_2018.1.163/linux/mkl/lib/intel64_lin/locale/%l_%t/%N:/usr/local/intel/parallel_studio_xe_2018/debugger_2018/gdb/intel64/share/locale/%l_%t/%N 
-------------------------------------------------------------------
```

### LINKED ISSUE #12684: "Extracted path from module does not exist": Truncated path from module
I'm getting the same warning message, for which I may have found the root cause, even if the problem may be slightly different.

### Steps to reproduce the issue
I get a warning when I do `spack spec llvm@8.0.0`, which requires python. For the latter one, I changed `packages.yaml` to use a module provided by the system.

The interesting line inside the python module is
```console
prepend-path    LD_LIBRARY_PATH    /opt/cray/pe/gcc-libs
```

### Error Message
```console
==> Warning: Extracted path from module does not exist:
	Extracted path: /opt/cray/pe/gcc-lib
```

***The problem, in my case, is that the last character gets truncated.***

I saw that #12662, but I think that the issue is slightly different, even if the problem may reside in the same code portion.

### PR DESCRIPTION
**Pull request by becker33**
A couple of issues have come up recently with module parsing heuristics. This PR makes the heuristics more robust, and adds testing for the portions that failed.

Fixes #12650
Fixes #12662
Fixes #12684
Closes #12687
Fixes #9710
Fixes #9126 
Fixes #5273

### CODE REVIEW

**Comment by adamjstewart** [comment]
Can confirm that this fixes #12662

**Comment by adamjstewart** [comment]
See if you can fix #9710 as well. There are some additional module parsing bugs noted in #9154 too.

**Comment by mwkrentel** [comment]
And I can confirm that this also fixes #12650.

**Comment by becker33** [comment]
@adamjstewart This should already fix #9710 and the issues related to module parsing in #9154. I've asked the bug reporters for #9126 and #5273 to check whether it resolves those issues as well.

File: lib/spack/spack/util/module_cmd.py [hunk 1]
```diff
@@ -119,37 +121,80 @@ def get_path_from_module_contents(text, module_name):
...
+    def strip_path(path, endings):
+        for ending in endings:
+            if path.endswith(ending):
+                return path[:-len(ending)]
+            if path.endswith(ending + '/'):
+                return path[:-(len(ending) + 1)]
+        return path
+
     for line in text:
+        # Check entries of LD_LIBRARY_PATH and CRAY_LD_LIBRARY_PATH
```
Comment (tldahlgren) [inline lib/spack/spack/util/module_cmd.py]: The following three searches here and two more starting at line 181 seem ripe for a function that takes a `pattern` and optional list of "subpaths" to be stripped.

File: lib/spack/spack/util/module_cmd.py [hunk 2]
```diff
@@ -119,37 +121,80 @@ def get_path_from_module_contents(text, module_name):
...
+        # Check {name}_ROOT entries
+        pattern = r'\W{0}_ROOT'.format(pkg_var_prefix)
+        if re.search(pattern, line):
+            paths = get_path_args_from_module_line(line)
+            for path in paths:
+                path_occurrences[path] = path_occurrences.get(path, 0) + 1
+
+        # Check entries that add a `-rpath` flag to a variable
         rpath = line.find('-rpath/')
         if rpath >= 0:
```
Comment (tldahlgren) [inline lib/spack/spack/util/module_cmd.py]: Similarly, the following two checks could be cleaner using a function.

**Comment by becker33** [comment]
@tldahlgren I've addressed your requests.

File: lib/spack/spack/util/module_cmd.py [hunk 3]
```diff
@@ -119,37 +121,68 @@ def get_path_from_module_contents(text, module_name):
...
+        # Check entries that update the PATH variable
         pattern = r'\WPATH'
-        if re.search(pattern, line):
-            path = get_path_arg_from_module_line(line)
-            return path[:path.find('/bin')]
+        match_pattern_and_strip(line, pattern, bin_endings)
+
+        # Check entries that update the MANPATH variable
+        pattern = r'MANPATH'
+        match_pattern_and_strip(line, pattern, man_endings)
```
Comment (tldahlgren) [inline lib/spack/spack/util/module_cmd.py]: Nit:  Why use pattern variable?

File: lib/spack/spack/util/module_cmd.py [hunk 4]
```diff
@@ -119,37 +121,68 @@ def get_path_from_module_contents(text, module_name):
...
+        match_pattern_and_strip(line, pattern)
 
-    # If it sets the PATH, use it
-    for line in text:
+        # Check {name}_ROOT entries
+        pattern = r'\W{0}_ROOT'.format(pkg_var_prefix)
+        match_pattern_and_strip(line, pattern)
+
+        # Check entries that update the PATH variable
         pattern = r'\WPATH'
```
Comment (tldahlgren) [inline lib/spack/spack/util/module_cmd.py]: Nit:  Why use any (short) pattern variables here?

**Pull request merged**